### ISSUE TITLE
presenter view leaking into external monitor view

### ISSUE THREAD

**Issue by Niklatz** [pdfpc/pdfpc#68]
When starting pdfpc without any arguments I have a problem with the presenter view (window with current slide, next slide, notes etc.) leaking into the external monitor. This would mean, that during the presentation on the left side of the screen there is always part of the presenter view visible on the projector.

See this picture for an example:
![pdfpc_presenter_too_large](https://cloud.githubusercontent.com/assets/3473679/8660590/1c833130-29b2-11e5-8836-46d52c549a4e.png)

The laptop screen ends at the red line, everything beyond that is visible on the external screen.

Am I missing some settings I could use to define the size of the presenter window? If i run it with the -z option it forces the view for the external monitor into windowed mode. During a presentation that would constantly mean a titlebar above the slides.

Unfortunately I could not find the part of the code where the presenter screen size is defined. Some pointers would be helpful so I could try to fixing it myself.

I hope I am not just too stupid to use it correctly. Here some info on my setup in case that is helpful:

Arch Linux 4.0.7
Xfce 4.12
pdfpc 4.0
Laptop screen resolution (presenter screen): 1280x800
External monitor resolution: 1920x1080

**Comment by AndreasBilke** [comment]
Out of curiosity, what happens if you change the resolution of your
external monitor to fit into the same aspect ratio (like your monitor)?

**Comment by Niklatz** [comment]
Unfortunately that doesn't change anything.

I set the resolution of both monitors to the same value (1280x800) and started the presentation as usual (no command line switches) but the presenter view is still too big for the Laptops screen.

Here is another image. Again everything right of the red line is seen on the external monitor:
![pdfpc_same_resolution](https://cloud.githubusercontent.com/assets/3473679/8668245/44ff8f76-2a09-11e5-869a-2667b9c215f8.png)

**Comment by AndreasBilke** [comment]
Does this also happen with pdfpc v3.1?

**Comment by rschroll** [comment]
From your first screenshot, I notice that the presenter window is 
completely the wrong size -- too short and too wide.  This is very odd. 
 If you launch it with the --switch-screens option, what happens?  What 
about if you launch pdfpc without an external monitor?

The window sizing is done in src/classes/window/fullscreen.vala.  
There's a few different code paths, depending on the options you've 
chosen, but when the windows are full screen, the sizing is done by 
calling Gtk.Window.fullscreen().  I worry that there's some 
disagreement between Gtk and your window manager that's giving it the 
wrong size.

**Comment by Niklatz** [comment]
@AndreasBilke Unfortunately v3.1 does not compile for me. I checked out the git repository tag `v3.1`, ran git submodule init and update and tried running cmake and make but it fails with `package 'gee-1.0' not found`. Any tips how I could test it?

@rschroll I believe the odd window sizing in the first screen shot is due to my monitor layout. I aligned the bottoms of the two screens so the external monitor is higher than the laptops monitor:
![bildschirmfoto_2015-07-14_21-40-51](https://cloud.githubusercontent.com/assets/3473679/8683099/0a279516-2a71-11e5-9479-8ea9f8fc2339.png)
However changing the alignment doesn't seem to  make any difference. If I align the tops of the screens it still leaks into the external monitor. The visible parts are just higher in the external view.

Using the `--switch-screens` behaves as expected. The presentation is started in fullscreen on the laptop monitor with the presenter view being visible on the external monitor. No leaking and both screens are nicely filled. However on the laptop screen the xfce-panel is seen over the presentation view. If I click into the presentation view it fills the screen but clicking out of it brings the panel back on top again.

**Comment by AndreasBilke** [comment]
@Niklatz you can download a v3.1 release from here https://github.com/davvil/pdfpc/releases . There are pre-generated c-sources included and you don't requiere the whole vala stuff.

Just unzip it, change into the folder and do `cmake .` CMake is searching for gee-1.0 dependencies. If you do not have them, you can install `gee06` from AUR (I tested it myself). Than you can build pdfpc.

**Comment by Niklatz** [comment]
@AndreasBilke thank you I got v3.1 compiled and tested it.

In v3.1 there is no leaking between the screens. The presenter view nicely fills the laptops screen and the presentation is displayed fullscreen on the external monitor. For comparison another screenshot:
![v3_1_screenshot](https://cloud.githubusercontent.com/assets/3473679/8698492/0d0c70e2-2aff-11e5-910c-cb85984dbcab.png)

What would be the best way to find the source of the problem? I would be willing to run some tests if you are interested but don't know where to look or what to try.

**Comment by AndreasBilke** [comment]
Is this correct that you can see your desktop while pdfpc is running? Actually, this should not happen.

For the first try, you could check commit 36e0a96e4c832c7263b14f96bba01847d7456343 if the problem was introduced there. With this commit we switched to gtk 3. Maybe the leaking starts there.

If 36e0a96e4c832c7263b14f96bba01847d7456343 was not the problem, you can use git-bisect to find the faulty commit.

**Comment by Niklatz** [comment]
No I can't see my desktop background while pdfpc is running. I probably should have mentioned this before but the top left part of the screenshot is not visible for me. This seems to be something the xfce4-screenshooter does to get a rectangle image. The display ends at the top of the presenter view on the left side and both screens are completely filled with the correct windows.

As for commit `36e0a96`: The presenter view is not leaking into the external monitor. However there is some problem in the presenter view. Areas that were black in `v3.1` and `v4.0` are now grayish (not desktop background color, stays the same if background is changed):
![36e0a96](https://cloud.githubusercontent.com/assets/3473679/8699291/cfefeec2-2b04-11e5-896e-973ed194bb18.png)

Also there is some output coming to the terminal with GtK warnings:
`(pdfpc:8067): Gtk-WARNING **: Theme parsing error: gtk-widgets.css:2224:19: Theming engine 'adwaita' not found`
This seems to be related to my selected color theme (correct word? I use the german translation of Xfce and it is called "Erscheinungsbild").

To make sure this wasn't the initial problem I tested some default settings for the color theme and received fewer GtK warnings to my terminal. These remained even with a default theme (were also present with previous settings):
`Initializing rendering...
Warning: Could not load icon /usr/local/share/pixmaps/pdfpc/blank.svg (Vorgang wird nicht unterstützt)
Warning: Could not load icon /usr/local/share/pixmaps/pdfpc/snow.svg (Vorgang wird nicht unterstützt)
Warning: Could not load icon /usr/local/share/pixmaps/pdfpc/pause.svg (Vorgang wird nicht unterstützt)`
The gray color in the presenter view also changed slightly but did not disappear completely:
![36e0a96_default_themes](https://cloud.githubusercontent.com/assets/3473679/8699566/a6078050-2b06-11e5-8425-aa0be2a88158.png)

Note: changing the color theme did not have any effect on `v4.0`. Here the presenter view is still leaking over with the Xfce4 default settings.

**Comment by AndreasBilke** [comment]
> As for commit `36e0a96`: The presenter view is not leaking into the external monitor.

Great. So now its time for bisecting until we find the faulty commit. Some
color issues etc are not the problem, since there was a lot of progress
since commit 36e0a96 to tackle some things.

**Comment by Niklatz** [comment]
Alright first time for me to use git bisect so to make sure I didn't mess it up I will quickly describe what I did:

`git checkout master`
`git bisect start`
`git bisect bad`
`git bisect good 36e0a96`

Then I kept building the versions while git moved through the commits and checked the created executable with the demo-pdf. Everytime I saw the issue i used `git bisect bad` and deleted the build files to make sure the next version would be a new build.

After a few steps of this git says that the a5105be534897a79bdd674a4b066d4ba204db970
is the first bad commit.

**Comment by AndreasBilke** [comment]
Actually, it is not obvious for me why the switch to CSS layouting is a problem. Could you add the following code to window/Fullscreen.vala after line 63 (master) and copy the output?

```
stdout.printf("screen_num %d, x %d, y %d, width %d, height %d\n", screen_num,
    this.screen_geometry.x, this.screen_geometry.y, this.screen_geometry.width,
    this.screen_geometry.height);
```

**Comment by AndreasBilke** [comment]
Please also post the `xrandr` output.

**Comment by Niklatz** [comment]
Ouput of pdfpc with added printf statement (I left out the GtK-Warning messages caused by my theme):
`screen_num 0, x 0, y 280, width 1280, height 800`
`screen_num 1, x 1280, y 0, width 1920, height 1080`
The printed resultions match my selected screen resoltions.

Outpout of `xrandr`:
`Screen 0: minimum 8 x 8, current 3200 x 1080, maximum 32767 x 32767`
`LVDS1 connected primary 1280x800+0+280 (normal left inverted right x axis y axis) 261mm x 163mm`
`   1280x800      59.99*+  50.00  `
`   1024x768      60.00  `
`   800x600       60.32    56.25  `
`   640x480       59.94  `
`   640x400       60.00  `
`DP1 disconnected (normal left inverted right x axis y axis)`
`DP2 disconnected (normal left inverted right x axis y axis)`
`DP3 disconnected (normal left inverted right x axis y axis)`
`HDMI1 disconnected (normal left inverted right x axis y axis)`
`HDMI2 disconnected (normal left inverted right x axis y axis)`
`VGA1 connected 1920x1080+1280+0 (normal left inverted right x axis y axis) 480mm x 270mm`
`   1920x1080     60.00*+`
`   1680x1050     59.95  `
`   1400x1050     59.98  `
`   1600x900      59.98  `
`   1280x1024     75.02    60.02  `
`   1440x900      59.89  `
`   1280x800      59.81  `
`   1152x864      75.00  `
`   1280x720      59.97  `
`   1024x768      75.08    60.00  `
`   800x600       75.00    60.32  `
`   640x480       75.00    60.00  `
`   720x400       70.08  `
`VIRTUAL1 disconnected (normal left inverted right x axis y axis)`

**Comment by einhander** [comment]
Hi, developers.
I have the same issue with screen size.
I use awesome wm for window management and if I change tag, pdfpc window  became normal size.
I have special rule in my config for pdfpc, may be it's  force pdfpc's window turn to normal.

**Comment by AndreasBilke** [comment]
@Niklatz I found a way to reproduce this issue also for me. To check if we can "fix" this could you checkout my pdfpc branch issue-32 and edit the file `rc/pdfpc.css`. You could try to change the following css snipped
```
#prerenderProgress {
    font-size: 20px;
}
```
and reduce the font-size to something small and test if the window still leaks into the other screen?

The issue-32 branch has nothing to do with a actual fix but in that branch you can easily play around with font sizes :)

**Comment by Niklatz** [comment]
@AndreasBilke This helped a lot but it did unfortunately not remove the issue entirely. I tried the issue-32 branch with two different settings:
```
#prerenderProgress {
    font-size: 20px;
}
```
With this setting the leaking occured like described before. Screenshot:
![20px](https://cloud.githubusercontent.com/assets/3473679/10071792/e949db2c-62bc-11e5-926d-ff8104303d69.png)

Then as recommended I lowered the value to:
```
#prerenderProgress {
    font-size: 5px;
}
```
In that case the leaking clearly reduced but there is still a very small amount left (small corner in the left edge of the presentation view about 3/4 down). See this screenshot:
![5px](https://cloud.githubusercontent.com/assets/3473679/10071852/514555b2-62bd-11e5-960a-a07750b141fd.png)

Reducing the setting further (`font-size: 1px;`) did not make any difference.

*Note:*
While testing this I had to use a different (much older) external monitor. To make sure this didn't introduce any new problems I checked with the known good version `v3.1`. Here no corner is visible. For direct comparison see this screenshot:
![v3_1](https://cloud.githubusercontent.com/assets/3473679/10071970/3241c870-62be-11e5-86f2-f72ca0e4c4c6.png)

**Comment by AndreasBilke** [comment]
@Niklatz can you solve your issue with changing the font-size in the css file?

**Comment by Niklatz** [comment]
Well like I [said](https://github.com/pdfpc/pdfpc/issues/68#issuecomment-142900372) changing the font size of 
```
#prerenderProgress {
    font-size: 20px;
}
```
in the no longer existing (?) issue-32 branch to a lower value somewhat helpes but there is still a small bar left on the presenter display. So far I haven't been able to solve the issue.

Is there something new in the current master branch I could try?

**Comment by AndreasBilke** [comment]
Yes, the size of the countdown and the slide progress are configurable ...

**Comment by Niklatz** [comment]
Not perfect but it's getting closer I think. The setting I played around with was in file `pdfpc/rc/pdfpc.css`
```
.bottomText {
    font-size: 5em;
}
```
Is that the one you meant?

With the default size (`5em`) and smaller there is a very small (I would guess 1 px) strip of the "next slide preview" visible. However, clicking into the presentation window on the external monitor brings this one on top and the strip is no longer visible. The controls still work fine so it is usable like that.

Since a screenshot would not really help in showing what I mean, here is a quick picture I took with my phone. Quality is not ideal but I hope you get the point:
![2015-11-09 21 39 06](https://cloud.githubusercontent.com/assets/3473679/11045917/8e60f82a-872a-11e5-855d-b5efd7f406b9.jpg)
Reducing the font-size further (`font-size: 1em;`) did not make any difference as far as leaking goes. The progress bar and timer got small but the 1 px wide stripe remained.

Increasing the value (`font-size: 10em;`) leads to  very much of the presenter view leaking over.

Screenshot:
![bildschirmfoto_2015-11-09_21-41-32](https://cloud.githubusercontent.com/assets/3473679/11045966/c84898e0-872a-11e5-9850-a8490040a59d.png)

Let me know if there is anything else I can do and thanks a lot for your effort!

**Comment by einhander** [comment]
Hi again, this issue is fixed for me in pdfpc 4.0.1

**Comment by AndreasBilke** [comment]
@Niklatz Could you try the current master?

**Comment by Niklatz** [comment]
I tried the current master (d256998021869181f02b9a45bc8a184922f8f8c5) and the problem still seems to be the same.

Starting the demo presentation opens the presenter view with the current and next slide with timer and page number at the bottom, and the presentation view with just the current slide. However there is still that ~1px wide strip of the "next slide" preview from the presenter view visible on the external monitor showing the presentation.

Here is another screenshot. I positioned the cursor right against the edge of the external monitor so you have an idea how much the presenter view is leaking over (everything right of the very tip of the cursor is visble on the external monitor).
![d256998](https://cloud.githubusercontent.com/assets/3473679/12728680/4cf95716-c923-11e5-9a5d-00ecda76da64.png)

I also tested the previously mentioned setting `.bottomText font-size` in `pdfpc/rc/pdfpc.css` but it did not make a difference. Even with `font-size: 1em` the leaking still occured.

**Comment by andybarry** [comment]
I am able to reproduce this by changing my resolutions to match yours and having the smaller screen set to the left of the bigger screen.  However, with  that configuration  ````font-size: 1em```` solves the problem for me.  Are you sure the CSS file you're using is being read?

**Comment by Niklatz** [comment]
Yes I can clearly see the timer and the slide number getting smaller when I change that value.

**Comment by AndreasBilke** [comment]
@Niklatz I did recently some work on how the lower status bar is sized etc. Could you try the latest master?

**Comment by Niklatz** [comment]
Sorry for the delay. I tried the current master with my default settings (laptop resolution=`1280x800` and external monitor resolution=`1920x1080`) and everything looks great. No leaking over to the presentation view.

However I am very sorry to tell you that there still seems to be some issue. For test purposes I reduced the resolution on my laptop screen to `800x600` and could once again observe the leaking into the presentation view (basically like in the very first screenshot of this issue).

So as far as I am concerned the issue is solved for everyday use but the real cause is still hidden somewhere. Anyway I would like to thank you for the great tool you (and others) created and the work you put into making it even better!
